first pass on docs lessons (#47)
MolSSI-Education/python-package-best-practices: _episodes/09-documentation.md

~~~
{: .output}

In particular, notice the `_build` directory. 
Sphinx put the HTML files it generated here. 
There should now be files in your `_build/html` folder. You can open these files in your browser to preview what your documentation will look like.

## Adding Information to our Documentation

Now, we will have the task of actually adding information about our project to our documentation. 
This part will require us to write! 

> When writing Sphinx documentation, you use `reStructured Text (RST)` syntax. 

Reviewer: ```suggestion
When writing Sphinx documentation, you use an extended `reStructuredText (RST)` [syntax](https://www.sphinx-doc.org/en/master/usage/restructuredtext/index.html). 
```

> This is similar to markdown, but has notable difference. See this [reStructured Text Cheat Sheet](https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html) for info on reStructured Text.

Reviewer: ```suggestion
This is similar to Markdown, but has notable differences.
See this [reStructured Text Cheat Sheet](https://thomas-cokelaer.info/tutorials/sphinx/rest_syntax.html) for info on reStructuredText.
```

Refresh the page in your browser to see the page with the new description.

> CookieCutter has already added some other pages to our documentation, `getting_started.rst`, and `api.rst`. 

Reviewer: ```suggestion
CookieCutter has already added some other pages to our documentation: `getting_started.rst`, and `api.rst`. 
```

## Sphinx Directives

The thing that really gives Sphinx special abilities are **directives**. 

> Directives are extensions which can be used with your restructred text to add special sections, images, warnings, etc to your documentation, and are part of a package called [docutils](https://docutils.sourceforge.io/) which Sphinx is built on top of

Reviewer: ```suggestion
Directives are extensions which can be used with your restructred text to add special sections, images, warnings, etc to your documentation, and are part of a package called [docutils](https://docutils.sourceforge.io/), which Sphinx is built on top of.
```

   api
~~~

This directive has an example of directive options, namely `maxdepth` and `caption`. 

> You can think of these as arguments into a function. 

Reviewer: ```suggestion
You can think of these as arguments to a function. 
```

We then have options for depth of the the Table of Contents (what level of headers to show), as well as the "caption". 
You can see other options for this directive under "Additional options" on the [documentation page](https://www.sphinx-doc.org/en/1.8/usage/restructuredtext/directives.html). 

After the settings for the TOC tree, you list the name of the pages you want to be included in this Table of Contents. 
Right now, we are including the 'Getting Started' page and the API page.

Note that you add the __file name__ in the TOC Tree, but the title of the page is what shows up in the TOC.

> For example, for the `Getting Started` page, the page heading ('Getting Started') , but you list the file name without the extension is the TOC tree as `getting_started`.

Reviewer: ```suggestion
For example, for the `Getting Started` page, the page heading is 'Getting Started', but you list the file name without the extension is the TOC tree as `getting_started`.
```

> ## Exercise

> > Create a new page with the title "About" that gives a description of the project. Specify that this project is for a MolSSI Best Practices Workshop, and put a link to the MolSSI homepage (molssi.org). 

Reviewer: ```suggestion
> Create a new page with the title "About" that gives a description of the project.
> Specify that this project is for a MolSSI Best Practices Workshop, and put a link to the MolSSI homepage (molssi.org). 
```

> Add your new page to the Table of Contents.
>

> > What happens if you remove the page from the table of contents and rebuild the documentation.

Reviewer: ```suggestion
> What happens if you remove the page from the table of contents and rebuild the documentation?
```

.. autofunction:: calculate_angle
~~~

> ## Structuring your documentation

> > Realistically, you probably don't want just a single page for API documentation. You could separate your documentation by purpose. For example, you could have separate pages explaining visualization, calculation, and/or measurement, with an autosummary table of contents for each page that has relevant functions.

Reviewer: ```suggestion
> Realistically, you probably don't want just a single page for API documentation.
> You could separate your documentation by purpose.
> For example, you could have separate pages explaining visualization, calculation, and/or measurement, with an autosummary table of contents for each page that has relevant functions.
```

>

> > Many projects are structured this way. Now that you know how Python documentation is built, keep an eye out for documentation you like. 

Reviewer: ```suggestion
> Many projects are structured this way.
> Now that you know how Python documentation is built, keep an eye out for documentation you like. 
```

The `mathjax` extension allows us to render latex into equations, and the `viewcode` extensions will add links to highlighted source code.

Next, we have added the line `autosummary_generate = True` to allow us to pull auto summaries from our modules and functions.

## Automatically Generating Documentation Using Sphinx-AutoAPI
Sphinx comes bundled with some extensions which have other directives. 
You can also download and install many Sphinx extensions from pip or conda.

One common use of Sphinx is to create API documentation (ie, pull the docstrings from your modules).
The cookiecutter comes bundled with Sphinx AutoSummary and AutoDoc extensions. 
However, you may have noticed that these are a bit labor-intensive if all you want to do is pull out the docstrings. 
If this is the approach you would like to take, you can use a Sphinx extension called `AutoAPI`. 
AutoAPI will pull documentation for all of your functions at once, rather than you having to build them manually.

> This is [Spinx-AutoAPI](https://github.com/readthedocs/sphinx-autoapi). To install Sphinx-AutoAPI, use `pip`:

Reviewer: ```suggestion
This is [Spinx-AutoAPI](https://github.com/readthedocs/sphinx-autoapi).
To install Sphinx-AutoAPI, use `pip`:
```

{: .language-python}

Once compiled, this will generate documentation for all files in the `molecool` source directory, as well as exclude
generating documentation for the tests and versioneering files in the project. 
The `index.rst` file will be updated to include an `API Reference` page, and the resultant `.html` files can be viewed in the `_build` directory upon 
compilation.

## Hosting your documentation

### Read The Docs

> We recommend hosing your  documentation on [Read The Docs](https://readthedocs.org/). 

Reviewer: ```suggestion
We recommend hosting your  documentation on [Read The Docs](https://readthedocs.org/). 
```
:-)

With this service, you can enable the building of your documentation every time you push to your repository.

Go to the [Read the Docs](https://readthedocs.org/) website. 

> Log in with your GitHub username and hook the repository to read the docs. 

Reviewer: ```suggestion
Log in with your GitHub username and hook the repository to ReadTheDocs. 
```

> Push to the repository, or trigger a build on the site. Because of the recent switch from `master` to `main`, you may get an error that your build has failed. 

Reviewer: ```suggestion
Push to the repository, or trigger a build on the site.
Because of the recent switch from `master` to `main`, you may get an error that your build has failed. 
```
